mbtools/replace_im_links.py

```diff
@@ -7,6 +9,24 @@
 from mbtools.fetch_im_resources import IM_PREFIX
 
 
+def replace_src_values_tree(content_tree, src_content, swap_mapping):
+    num_changes = 0
+    for elem in utils.find_elements_containing(content_tree, src_content):
+        im_filename = elem.attrib["src"]
+        if im_filename in swap_mapping.keys():
+            num_changes += 1
+            elem.attrib["src"] = swap_mapping[im_filename]
+        else:
+            # Fetch link, calculate sha1, and replace link anyway.
+            data = requests.get(im_filename).content
+            sha1 = hashlib.sha1(data).hexdigest()
+            for key in swap_mapping.keys():
```

Comment:
I didn't immediately realize this, but it looks like you're iterating here making this lookup `O(n)`. It also makes an implicit assumption that the target URL includes the sha1 value. Would it be worthwhile to perhaps have `parse_media_file` return two mapping objects?:

1. The existing IM URL -> `entry['s3_url']`
2. A new `entry['sha1']` -> `entry['s3_url']`

You could then pass both mappings into this function and make this branch a bit more efficient and perhaps improve clarity :man_shrugging: .